test/upgrade/types.go

```diff
@@ -84,10 +86,12 @@ type Context struct {
 // down to each handler of BackgroundOperation. It contains a StopEvent channel
 // which end user should use to obtain a testing.T for error reporting. Until
 // StopEvent is sent user may use zap.SugaredLogger to log state of execution if
-// necessary.
+// necessary. The logs are stored in a ThreadSafeBuffer and flushed to the test
+// output when the test fails.
 type BackgroundContext struct {
-	Log  *zap.SugaredLogger
-	Stop <-chan StopEvent
+	Log       *zap.SugaredLogger
+	LogBuffer *ThreadSafeBuffer
```

Comment:
I don't like adding this internal thing to public struct as public field. As minimal effort, to should be `logBuffer` if this needs to be here at all.